src/java/org/apache/cassandra/db/compaction/CompactionStrategyManager.java

```diff
@@ -702,6 +793,85 @@ public boolean isLeveledCompaction()
         }
     }
 
+    double[] avgUCSHelper(BiFunction<UnifiedCompactionStrategy, SSTableReader, Double> fn)
+    {
+        readLock.lock();
+        try
+        {
+            if (repaired.first() instanceof UnifiedCompactionStrategy)
+            {
+                int numberOfLevels = 0;
+
+                double[] sum = new double[UnifiedCompactionStrategy.MAX_LEVELS];
+                int[] count = new int[UnifiedCompactionStrategy.MAX_LEVELS];
+
+                for (AbstractCompactionStrategy strategy : getAllStrategies())
+                {
+                    UnifiedCompactionStrategy ucsStrategy = (UnifiedCompactionStrategy) strategy;
+                    List<Level> levels = ucsStrategy.getLevelsSnapshot();
+                    numberOfLevels = Math.max(numberOfLevels, levels.size());
+                    for (int i = 0; i < levels.size(); i++)
+                    {
+                        for (SSTableReader sstable : levels.get(i).getSSTables())
+                        {
+                            sum[i] += fn.apply(ucsStrategy, sstable);
+                            count[i] += 1;
+                        }
+                    }
+                }
+
+                double[] res = new double[numberOfLevels];
+                for (int i = 0; i < numberOfLevels; i++)
+                    res[i] = count[i] == 0 ? 0 : sum[i] / count[i];
+
+                return res;
+            }
+            return null;
+        }
+        finally
+        {
+            readLock.unlock();
+        }
+    }
+
+    double[] perLevelUCSHelper(TriFunction<Level, SSTableReader, Double, Double> fn)
+    {
+        readLock.lock();
+        try
+        {
+            if (repaired.first() instanceof UnifiedCompactionStrategy)
+            {
+                int numberOfLevels = 0;
+
+                double[] tmp = new double[UnifiedCompactionStrategy.MAX_LEVELS];
+
+                for (AbstractCompactionStrategy strategy : getAllStrategies())
+                {
+                    UnifiedCompactionStrategy ucsStrategy = (UnifiedCompactionStrategy) strategy;
+                    List<Level> levels = ucsStrategy.getLevelsSnapshot();
+                    numberOfLevels = Math.max(numberOfLevels, levels.size());
+                    for (int i = 0; i < levels.size(); i++)
+                    {
+                        for (SSTableReader sstable : levels.get(i).getSSTables())
+                        {
+                            tmp[i] = fn.apply(levels.get(i), sstable, tmp[i]);
+                        }
+                    }
+                }
+
+                double[] res = new double[numberOfLevels];
+                System.arraycopy(tmp, 0, res, 0, numberOfLevels);
+
+                return res;
+            }
+            return null;
+        }
+        finally
+        {
+            readLock.unlock();
+        }
+    }
+
```

Comment:
I noticed `getPerLevelAvgDensityMaxDensityThresholdRatio(), getPerLevelMaxDensityMaxDensityThresholdRatio(), avgUCSHelper(), and perLevelUCSHelper() `all share common code patterns (lock acquisition, iterating across levels and sstables, result array creation).               

Suggestion: We could consolidate these into a single generic helper:

computeUCSMetric(Consumer<UCSLevelData> accumulator, IntFunction<Double> finalizer)

  Where:
  - accumulator contains the per-sstable operation logic (what to do in the loop)
  - finalizer computes the final result for each level

  This would eliminate the duplication while keeping each metric method focused on its specific calculation logic.

  Would this approach work for you? Happy to provide a more detailed example if helpful!